dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/icon/DefaultIconService.java

```diff
@@ -69,16 +77,51 @@ public class DefaultIconService implements IconService {
 
   private final UserService userService;
 
+  private final JobCreationHelper jobCreationHelper;
+
   @Override
   @Transactional
   public void createDefaultIcons() {
-    // build a set of all default icon variant combination that we want to have exiting
-    // build a diff with DB state => missing icons (select iconkey from icon where iconkey not in
-    // (%placeholder);)
-    // diff == nothing? => return;
-    // loop missing icons for insert =>
-    // 1. create FileResource for InputStream (use syncSaveFileResource) see JobCreationHelper
-    // 2. call addIconInternal
+    List<String> keysNotInDatabase =
+        CollectionUtils.difference(getInMemoryKeysWithVariants(), iconStore.getIconKeys());
+
+    if (keysNotInDatabase.isEmpty()) return;
+
+    List<Icon> icons =
+        Arrays.stream(DefaultIcon.values())
+            .map(this::createIconWithVariants)
+            .flatMap(Collection::stream)
+            .filter(i -> keysNotInDatabase.contains(i.getKey()))
+            .toList();
+
+    icons.forEach(
+        i -> {
+          try {
+
+            String fileResourceId = CodeGenerator.generateUid();
+
+            Resource resource = getDefaultIconResource(i.getKey());
+
+            jobCreationHelper.saveJobData(
+                fileResourceId,
+                MediaType.IMAGE_PNG,
+                resource.getInputStream(),
+                fileResourceService,
+                CUSTOM_ICON);
+
+            FileResource fileResource = fileResourceService.getFileResource(fileResourceId);
+
+            i.setFileResource(fileResource);
+
+            addIconInternal(i);
+          } catch (NotFoundException
+              | ConflictException
+              | IOException
+              | SQLException
+              | BadRequestException e) {
+            e.printStackTrace();
```

Comment:
Please log.error instead of printing the stack trace.

I am not sure when printing the stack trace like this would be appropriate. You want users to be able to configure what and how things get logged. This way they cannot.

Reply:
i think this code is outdated.